components/scream/src/physics/shoc/shoc_energy_fixer_impl.hpp

```diff
@@ -0,0 +1,99 @@
+#ifndef SHOC_ENERGY_FIXER_IMPL_HPP
+#define SHOC_ENERGY_FIXER_IMPL_HPP
+
+#include "shoc_functions.hpp" // for ETI only but harmless for GPU
+
+namespace scream {
+namespace shoc {
+
+/*
+ * Implementation of shoc shoc_energy_fixer. Clients should NOT
+ * #include this file, but include shoc_functions.hpp instead.
+ */
+
+template<typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>::shoc_energy_fixer(
+  const MemberType&            team,
+  const Int&                   nlev,
+  const Int&                   nlevi,
+  const Scalar&                dtime,
+  const Int&                   nadv,
+  const uview_1d<const Spack>& zt_grid,
+  const uview_1d<const Spack>& zi_grid,
+  const Scalar&                se_b,
+  const Scalar&                ke_b,
+  const Scalar&                wv_b,
+  const Scalar&                wl_b,
+  const Scalar&                se_a,
+  const Scalar&                ke_a,
+  const Scalar&                wv_a,
+  const Scalar&                wl_a,
+  const Scalar&                wthl_sfc,
+  const Scalar&                wqw_sfc,
+  const uview_1d<const Spack>& rho_zt,
+  const uview_1d<const Spack>& tke,
+  const uview_1d<const Spack>& pint,
+  const uview_1d<Spack>&       rho_zi,
+  const uview_1d<Spack>&       host_dse)
+{
+  // Constants
+  const auto cp = C::CP;
+  const auto lcond = C::LatVap;
+  const auto lice = C::LatIce;
+  const auto mintke = SC::mintke;
+  const auto ggr = C::gravit;
+
+  // Local variables
+  Scalar te_a = 0;
+  Scalar te_b = 0;
+  Scalar se_dis = 0;
+  Int shoctop = 0;
+
+  // Compute linear interpolation of data into rho_zi
+  linear_interp(team,zt_grid,zi_grid,rho_zt,rho_zi,nlev,nlevi,0);
+  team.team_barrier();
+
+  // Compute the host timestep
+  const Scalar hdtime = dtime*nadv;
+
+  const auto nlevi_pack = ekat::npack<Spack>(nlevi)-1;
+  const int nlevi_indx = (nlevi-1)%Spack::n;
+
+  // Compute the total energy before and after SHOC call
+  const Scalar shf = wthl_sfc*cp*rho_zi(nlevi_pack)[nlevi_indx];
+  const Scalar lhf = wqw_sfc*rho_zi(nlevi_pack)[nlevi_indx];
+  te_a = se_a + ke_a + (lcond+lice)*wv_a +lice*wl_a;
+  te_b = se_b + ke_b + (lcond+lice)*wv_b + lice*wl_b;
+  te_b += (shf+lhf*(lcond+lice))*hdtime;
+
+  //Find which pack and pack index corresponds to shoctop
+  int shoctop_pack = shoctop/Spack::n;
+  int shoctop_indx = shoctop%Spack::n;
+
+  // Limit the energy fixer to find highest layer where SHOC is active.
```

Comment:
This is technically correct since these are all local variables, but you can probably work out a better parallel scheme using a reduction over an integer, the level index.

Reply:
@ambrad Let me know what you think about this. Maybe a reduction over scalar index instead of pack index would be better?